README.md
# Alternative Starts Updated
Update for [Alternative Start](https://mods.factorio.com/mod/AlternativeStart) mod by *Vas*.
This mod gives you different options when starting a new map and joining the game. You can choose among different kits, each one with different starting items. 

### Mining kit

- **Base**: burner technology
- **Light Industry**: first electric drills
- **Medium Industry**: some electric drill and steel furnaces
- **Heavy Industry**: more electric drills
- **Tycoon**: set of electric drills and furnaces

### Military kit

- **Civilian**: pistol + ammunition
- **Soldier**: submachine gun and shotgun + ammunition
- **Rambo**: submachine, shotgun and flamethrower + ammunition

### Power kit

- **Steam**: steam engine
- **Advanced**: steam + solar support
- **Solar**: big solar farm
- **Solar Field**: huge solar farm

### Booster kit

- **Resources**: intermediates
- **Power Armor**: power armor + equipments
- **Research Equipment**: science packs
- **Power Armor & Resources**
- **All**

### Modules kit

- **Poker Speed**: 4x T3 Speed modules
- **Poker Productivity**: 4x T3 Productivity modules
- **Poker Efficiency**: 4x T4 Efficiency modules
- **Tris of Poker**: All 3 pokers
- **Royal Scale**: 4x Tris of Poker

### Robot kit

- **Base**: roboport + 20 mixed bots
- **Pack**: roboport + 50 mixed bots
- **Flock**: 2 roboports + 100 mixed bots
- **Fleet**: 4 roboports + 200 mixed bots

It is also possible to add logistic chests to enhance the robot kit.

### Vehicle kit

- **Car**: car + fuel
- **Tank**: tank + fuel
- **Train**: trains, wagons, rails & stations

### Landfill Kit

- **Small**: 50 landfill
- **Medium**: 100 landfill
- **Large**: 500 landfill
- **Huge**: 1000 landfill

---

## Translation
Available languages:

- English (en)

You can help by translating this mod into your language. Any translations made will be included in the next release.

---

## Known Bugs & Mod compatibility
Please feel free to report any known issue! :)

**Alternative Starts Updated** is developed for vanilla games and should be compatible with all mods that do not directly interfere with it. Requests for compatibility with other mods can be submitted of course! :)

---

## Future developments
- [ ] Re-establish compatibility with
  - AAI Vehicles
  - AAI Programmable Vehicles
  - Timed Technology
- [ ] implement custom starting items

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.6.3
Date: 11/03/2023
  Changes:
    - Updated config items
---------------------------------------------------------------------------------------------------
Version: 0.6.2
Date: 05/11/2022
  Changes:
    - Added Jedi configuration
---------------------------------------------------------------------------------------------------
Version: 0.6.1
Date: 17/12/2021
  Fixes:
    - Fixed crash on loading new games with no settings selected
    - Implemented additional check for default settings
---------------------------------------------------------------------------------------------------
Version: 0.6.0
Date: 25/11/2021
  Changes:
    - Updated to version 1.1
---------------------------------------------------------------------------------------------------
Version: 0.5.3
Date: 11/08/2019
  Changes:
    - Updated mod number to make work with 0.18 because devs refuse to allow mods with incorrect version to be loaded anyway.
---------------------------------------------------------------------------------------------------
Version: 0.5.2
Date: 11/08/2019
  Fixes:
    - Crash on load from line 270.
  Removed:
    - If statement checking for AAI mod, to fix crash on line 270.
---------------------------------------------------------------------------------------------------
Version: 0.5.1
Date: 11/08/2019
  Fixes:
    - Crash on load from line 227 resulting in nil.
  Removed:
    - if statement checking for bob's mods before letting you use modular port option. Can't make this work right now, resulted in crash.
---------------------------------------------------------------------------------------------------
Version: 0.5.0
Date: 11/07/2019
  Changes:
    - Removed axe from start.
    - Renamed science packs because the devs changed the code name for no particular reason other than to break all mods that used them.
    - Added modular robo port option.
  New Features:
    - Input Box for additional items not in the dropdowns.
    - Placeholder input box for Technologies to start with.
    - Placeholder input box for items to disable.
    - Placeholder input box for technologies to disable.
  Removed:
    - Money related events, since its not used right now.
---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 01/30/2018
  Bugfixes:
    - Fixed incorrect display order.
  Changes:
    - Renamed Shamefully Easy Electric to Mega Electric.
    - Added "Epic Electric" option that adds even more electric things and even gives you electric furnaces. This option also gives substations and uses substation multiplier.
    - Added blue science packs to lab start conditions.
    - Renamed Power Armor option to remove "modules" from the name so I can use it elsewhere.
    - Added a new modules setting to allow you to start with one of three types of modules or even all of them.
    - Renamed train option.
    - Altered the money option to no longer give you a static amount, the new amount given is based on what options you chose, which can drive you into the negative. In which case you'll simply get nothing.
---------------------------------------------------------------------------------------------------
Version: 0.2.5
Date: 01/26/2018
  Changes:
    - Added a beta train option, let me know if you believe I should increase or decrease what the option gives you in train parts.
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 01/21/2018
  Bugfixes:
    - Added missing stone furnace to electric drill option.
  Changes:
    - Changed two options from Boolean to Dropdown to give additional choices.
    - Improved and optimized the script a little to allow better inventory management of your choices.
    - New options added for weapon starting.
    - Changed some options from Boolean to Dropdown, allowing different levels of choices.
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 01/19/2018
  Bugfixes:
    - Fixed an issue with the mod truncicating items in the hotbar, now all items properly go to the correct locations except a few things which are for the moment placed inside the inventory instead of the hotbar.
---------------------------------------------------------------------------------------------------
Version: 0.0.1
Date: 01/19/2018
  Changes:
    - Created Mod


utils.lua
-- @ tableIn: table<{number, string}>
-- @ item: string
function RemoveItem(tableIn, item)
  if item ~= nil then
    if tableIn ~= nil then
      for index, value in pairs(tableIn) do
        if value[2] ~= nil then
          if value[2] == item then
            tableIn[index] = nil
          end
        end
      end
    end
  end
end

-- @ config: AS.config
-- @ item: string
function RemoveAllItemOccurrences(config, item)
  if config ~= nil then
    for config_key, config_values in pairs(config) do
      for kit_name, kit_option in pairs(config_values) do
        RemoveItem(kit_option, item)
      end
    end
  end
end

-- @ tableIn: table<{number, string}>
-- @ tuple: {number, string}
function AddItem(tableIn, tuple)
  if tuple ~= nil then
    if tableIn ~= nil then
      table.insert(tableIn, tuple)
    end
  end
end



control.lua
require "config"

AS.item_list = {}
AS.multiplier = (settings.global["basic-kit-substations"].value and 1 or 0) * AS.multiplier
AS.chests = settings.global["basic-kit-logistic-chests"].value and 1 or 0

-- parse settings values using config file fields
for key, options in pairs(AS.config) do
  if settings.global[key] ~= nil then
    local option_name = settings.global[key].value
    if options[option_name] ~= nil then
      for _, tuple in pairs(options[option_name]) do
        table.insert(AS.item_list, tuple)
      end
    else
      if settings.global[key].value ~= "None" then
        log("Options not defined for key: " .. option_name)
      end
    end
  else
    log("Undefined global setting with key: " .. key)
  end
end

-- add items to player inventory
local function InsertIntoInventory(event)
  local player = game.players[event.player_index]

	pcall(player.exit_cutscene)
	
	for _, item in pairs(AS.item_list) do
		if game.item_prototypes[item[2]] ~= nil then
			player.insert{ name = item[2], count = item[1]}
		else
			player.print("Unable to add \"" .. item[2] .. "\" to inventory, please check spelling.")
		end
	end
end

-- init compatibility
-- moved to config

-- add items to inventory
script.on_event(defines.events.on_player_created, InsertIntoInventory)

settings.lua
--[[
	Settings
	They are all runtime global because server sync
]]--

data:extend({
	-- Mining equipment
	{
		type = "string-setting",
		name = "basic-kit-mining",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Base",
			"Light Industry",
			"Medium Industry",
			"Heavy Industry",
			"Tycoon"
		},
		order = "as-100"
	},
	-- Military equipment
	{
		type = "string-setting",
		name = "basic-kit-military",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Civilian",
			"Soldier",
			"Rambo",
			"Jedi"
		},
		order = "as-200"
	},
	-- Power equipment
	{
		type = "string-setting",
		name = "basic-kit-power",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Steam",
			"Advanced",
			"Solar",
			"Solar Field",
		},
		order = "as-300"
	},
	-- Booster kit
	{
		type = "string-setting",
		name = "basic-kit-booster",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Resources",
			"Power Armor",
			"Research Equipment",
			"Power Armor & Resources",
			"All"
		},
		order = "as-400"
	},
	-- Modules kit
	{
		type = "string-setting",
		name = "basic-kit-modules",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Poker Speed",
			"Poker Productivity",
			"Poker Efficiency",
			"Tris of Poker",
			"Royal Scale"
		},
		order = "as-500"
	},
	-- Robots kit
	{
		type = "string-setting",
		name = "basic-kit-robot",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Base",
			"Pack",
			"Flock",
			"Fleet"
		},
		order = "as-600"
	},
	-- Veichle kit
	{
		type = "string-setting",
		name = "basic-kit-vehicle",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Car",
			"Tank",
			"Train"
		},
		order = "as-700"
	},
	-- Landfill kit
	{
		type = "string-setting",
		name = "basic-kit-landfill",
		setting_type = "runtime-global",
		default_value = "None",
		allowed_values = {
			"None",
			"Small",
			"Medium",
			"Large",
			"Huge"
		},
		order = "as-800"
	},
	-- Logistic chest option
	{
		type = "bool-setting",
		name = "basic-kit-logistic-chests",
		setting_type = "runtime-global",
		default_value = false,
		order = "as-650"
	},
	-- Additional substation option
	{
		type = "bool-setting",
		name = "basic-kit-substations",
		setting_type = "runtime-global",
		default_value = false,
		order = "as-350"
	}
})

--[[
	,
	{
		-- Choose the exact equipment you wish to start with.
		type = "string-setting",
		name = "basic-kit-start-string",
		localised_name = "Choose Extra Equipment",
		localised_description = "Here, you can choose the equipment you would like to start with separating them out with commas and giving an amount you wish to have. Referrer to the example given in default value. You can also get some preset values from the mod's upload page.",
		setting_type = "runtime-global",
		allow_blank = true,
		default_value = "iron-plate:4,iron-axe:1,burner-mining-drill:1,stone-furnace:1,pistol:1,firearm-magazine:10",
		order = "as-990"
	}
]]


archive/configurations.lua
--[[
	Code originally referenced from AyrA.
	Modified greatly and improved a lot by Vas.
	If you use this, please make sure to mention us somewhere like in your readme.
]]--

-- User Settings
local BASE_KIT				=	settings.global["basic-kit-base"].value
local EQUIPME					= settings.global["basic-kit-equipment"].value
local POWER_KIT				= settings.global["basic-kit-power"].value
local MORE_SUBS				= settings.global["basic-kit-moresubs"].value
local ADD_EASY				=	settings.global["basic-kit-easy"].value
local ADD_MODULES			= settings.global["basic-kit-module"].value
local ADD_ROBOTS			= settings.global["basic-kit-robot"].value
local ADD_ROCHEST			= settings.global["basic-kit-robotchests"].value
local ADD_CAR					= settings.global["basic-kit-car"].value
local PERFECTIONIST		=	settings.global["basic-kit-perfectionist"].value
local MODU_ROBO				= settings.global["basic-kit-modularrobots"].value
local START_EQUIP			= settings.global["basic-kit-start-string"].value
local START_TECH			=	settings.global["basic-kit-start-research-string"].value
local DISASABLE_EQUIP =	settings.global["basic-kit-disable-equipment-string"].value
local DISABLE_TECH		= settings.global["basic-kit-disable-research-string"].value

--Start the kit table.
local kit = {
	{ name = "iron-plate", count = 4}
}
-- Add stuff to inventory depending on the settings

-- Lets create a splitter function so we can split our additional strings properly.
local function split(inputstr, sep)
	if sep == nil then sep = "%s" end
	t={} i=1
	for str in string.gmatch(inputstr, "([^"..sep.."]+)") do 
		t[i] = str
		i = i + 1
	end return t
end

--[[Choose Mining Equipment]]--
if BASE_KIT == "Base" then
	table.insert(kit,{name="burner-mining-drill",count=1})
	table.insert(kit,{name="stone-furnace",count=1})
elseif BASE_KIT == "Electric" then
	table.insert(kit,{name="electric-mining-drill",count=1})
	table.insert(kit,{name="medium-electric-pole",count=1})
	table.insert(kit,{name="stone-furnace",count=1})
elseif BASE_KIT == "Easy Electric" then
	table.insert(kit,{name="electric-mining-drill",count=2})
	table.insert(kit,{name="medium-electric-pole",count=5})
	table.insert(kit,{name="stone-furnace",count=3})
elseif BASE_KIT == "Mega Electric" then
	table.insert(kit,{name="electric-mining-drill",count=4})
	table.insert(kit,{name="medium-electric-pole",count=20})
	table.insert(kit,{name="stone-furnace",count=6})
elseif BASE_KIT == "Epic Electric" then
	table.insert(kit,{name="electric-mining-drill",count=8})
	table.insert(kit,{name="electric-furnace",count=10})
	if MORE_SUBS then
		table.insert(kit,{name="substation",count=16})
	else
		table.insert(kit,{name="substation",count=4})
	end
end

--[[Choose your combat starting gear]]--
if EQUIPME == "Civilian" then
	table.insert(kit,{name="pistol",count=1})
	table.insert(kit,{name="firearm-magazine",count=10})
elseif EQUIPME == "Soldier" then
	table.insert(kit,{name="submachine-gun",count=1})
	table.insert(kit,{name="shotgun",count=1})
	table.insert(kit,{name="piercing-rounds-magazine",count=20})
	table.insert(kit,{name="piercing-shotgun-shell",count=20})
elseif EQUIPME == "Rambo" then
	table.insert(kit,{name="submachine-gun",count=1})
	table.insert(kit,{name="flamethrower",count=1})
	table.insert(kit,{name="combat-shotgun",count=1})
	table.insert(kit,{name="uranium-rounds-magazine",count=100})
	table.insert(kit,{name="flamethrower-ammo",count=50})
	table.insert(kit,{name="piercing-shotgun-shell",count=50})
end

--[[Steam or Solar start]]--
if POWER_KIT == "Steam" then
	table.insert(kit,{name="offshore-pump",count=1})
	table.insert(kit,{name="boiler",count=1})
	table.insert(kit,{name="pipe",count=10})
	table.insert(kit,{name="steam-engine",count=1})
elseif POWER_KIT == "Solar" then
	table.insert(kit,{name="solar-panel",count=6})
	table.insert(kit,{name="accumulator",count=3})
	if MORE_SUBS then
		table.insert(kit,{name="substation",count=8})
	else
		table.insert(kit,{name="substation",count=2})
	end
elseif POWER_KIT == "Super Solar" then
	table.insert(kit,{name="solar-panel",count=12})
	table.insert(kit,{name="accumulator",count=6})
	if MORE_SUBS then
		table.insert(kit,{name="substation",count=16})
	else
		table.insert(kit,{name="substation",count=4})
	end
elseif POWER_KIT == "Insane Solar" then
	table.insert(kit,{name="solar-panel",count=24})
	table.insert(kit,{name="accumulator",count=12})
	if MORE_SUBS then
		table.insert(kit,{name="substation",count=32})
	else
		table.insert(kit,{name="substation",count=8})
	end
end

--[[Easier Start]]--
if ADD_EASY == "Resources" then
	table.insert(kit,{name="iron-plate",count=46})
	table.insert(kit,{name="copper-plate",count=50})
elseif ADD_EASY == "Power Armor" then
	table.insert(kit,{name="modular-armor",count=1})
	table.insert(kit,{name="energy-shield-equipment",count=1})
	table.insert(kit,{name="night-vision-equipment",count=1})
	table.insert(kit,{name="battery-equipment",count=1})
	table.insert(kit,{name="solar-panel-equipment",count=7})
	table.insert(kit,{name="exoskeleton-equipment",count=1})
elseif ADD_EASY == "Research Equipment" then
	if game.active_mods["timed_technology"] then
		table.insert(kit,{name="lab",count=1})
	else
		table.insert(kit,{name="lab",count=1})
		table.insert(kit,{name="automation-science-pack",count=50}) --science-pack-1
		table.insert(kit,{name="logistic-science-pack",count=50}) --science-pack-2
		table.insert(kit,{name="chemical-science-pack",count=50}) --science-pack-3
	end
elseif ADD_EASY == "Power Armor & Resources" then
	table.insert(kit,{name="iron-plate",count=46})
	table.insert(kit,{name="copper-plate",count=50})
	table.insert(kit,{name="modular-armor",count=1})
	table.insert(kit,{name="energy-shield-equipment",count=1})
	table.insert(kit,{name="night-vision-equipment",count=1})
	table.insert(kit,{name="battery-equipment",count=1})
	table.insert(kit,{name="solar-panel-equipment",count=7})
	table.insert(kit,{name="exoskeleton-equipment",count=1})
elseif ADD_EASY == "All" then
	table.insert(kit,{name="iron-plate",count=46})
	table.insert(kit,{name="copper-plate",count=50})
	table.insert(kit,{name="modular-armor",count=1})
	table.insert(kit,{name="energy-shield-equipment",count=1})
	table.insert(kit,{name="night-vision-equipment",count=1})
	table.insert(kit,{name="battery-equipment",count=1})
	table.insert(kit,{name="solar-panel-equipment",count=7})
	table.insert(kit,{name="exoskeleton-equipment",count=1})
	table.insert(kit,{name="lab",count=1})
	table.insert(kit,{name="automation-science-pack",count=50}) --science-pack-1
	table.insert(kit,{name="logistic-science-pack",count=50}) --science-pack-2
	table.insert(kit,{name="chemical-science-pack",count=50}) --science-pack-3
end

--[[Modules]]--
if ADD_MODULES == "Speed" then
	table.insert(kit,{name="speed-module-3",count=4})
elseif ADD_MODULES == "Productivity" then
	table.insert(kit,{name="productivity-module-3",count=4})
elseif ADD_MODULES == "Energy Efficiency" then
	table.insert(kit,{name="effectivity-module-3",count=4})
elseif ADD_MODULES == "All" then
	table.insert(kit,{name="speed-module-3",count=4})
	table.insert(kit,{name="productivity-module-3",count=4})
	table.insert(kit,{name="effectivity-module-3",count=4})
elseif ADD_MODULES == "4x All" then
	table.insert(kit,{name="speed-module-3",count=16})
	table.insert(kit,{name="productivity-module-3",count=16})
	table.insert(kit,{name="effectivity-module-3",count=16})
end

--[[Start with bots]]--
if ADD_ROBOTS == "Basic" then
	if MODU_ROBO then --[[settings.global["bobmods-logistics-disableroboports"].value == true OR game.active_mods["boblogistics"]]--
		table.insert(kit,{name="bob-robochest",count=1})
		table.insert(kit,{name="bob-robo-charge-port",count=1})
		table.insert(kit,{name="bob-logistic-zone-interface",count=1})
		table.insert(kit,{name="bob-logistic-zone-expander",count=1})
	else
		table.insert(kit,{name="roboport",count=1})
	end
	table.insert(kit,{name="logistic-robot",count=10})
	table.insert(kit,{name="construction-robot",count=2})
	if ADD_ROCHEST then
		table.insert(kit,{name="logistic-chest-passive-provider",count=3})
		table.insert(kit,{name="logistic-chest-storage",count=1})
		table.insert(kit,{name="logistic-chest-requester",count=3})
	end
elseif ADD_ROBOTS == "Lots" then
	if MODU_ROBO then
		table.insert(kit,{name="bob-robochest",count=1})
		table.insert(kit,{name="bob-robo-charge-port",count=2})
		table.insert(kit,{name="bob-logistic-zone-interface",count=1})
		table.insert(kit,{name="bob-logistic-zone-expander-2",count=1})
	else
		table.insert(kit,{name="roboport",count=1})
	end
	table.insert(kit,{name="logistic-robot",count=25})
	table.insert(kit,{name="construction-robot",count=5})
	if ADD_ROCHEST then
		table.insert(kit,{name="logistic-chest-passive-provider",count=6})
		table.insert(kit,{name="logistic-chest-storage",count=3})
		table.insert(kit,{name="logistic-chest-requester",count=6})
	end
elseif ADD_ROBOTS == "Too Many" then
	if MODU_ROBO then
		table.insert(kit,{name="bob-robochest",count=1})
		table.insert(kit,{name="bob-robo-charge-port",count=3})
		table.insert(kit,{name="bob-logistic-zone-interface",count=1})
		table.insert(kit,{name="bob-logistic-zone-expander-3",count=2})
	else
		table.insert(kit,{name="roboport",count=2})
	end
	table.insert(kit,{name="logistic-robot",count=50})
	table.insert(kit,{name="construction-robot",count=10})
	if ADD_ROCHEST then
		table.insert(kit,{name="logistic-chest-passive-provider",count=10})
		table.insert(kit,{name="logistic-chest-storage",count=5})
		table.insert(kit,{name="logistic-chest-requester",count=10})
	end
elseif ADD_ROBOTS == "Too Many ++" then
	if MODU_ROBO then
		table.insert(kit,{name="bob-robochest",count=2})
		table.insert(kit,{name="bob-robo-charge-port",count=4})
		table.insert(kit,{name="bob-robo-charge-port-large",count=1})
		table.insert(kit,{name="bob-logistic-zone-interface",count=1})
		table.insert(kit,{name="bob-logistic-zone-expander-4",count=2})
	else
		table.insert(kit,{name="roboport",count=4})
	end
	table.insert(kit,{name="logistic-robot",count=100})
	table.insert(kit,{name="construction-robot",count=20})
	if ADD_ROCHEST then
		table.insert(kit,{name="logistic-chest-passive-provider",count=15})
		table.insert(kit,{name="logistic-chest-storage",count=10})
		table.insert(kit,{name="logistic-chest-requester",count=15})
		table.insert(kit,{name="logistic-chest-active-provider",count=5})
		table.insert(kit,{name="logistic-chest-buffer",count=5})
	end
end

--[[Add a car]]--
if ADD_CAR == "Car" then
	table.insert(kit,{name="car",count=1})
	table.insert(kit,{name="solid-fuel",count=5})
elseif ADD_CAR == "Tank" then
	table.insert(kit,{name="tank",count=1})
	table.insert(kit,{name="solid-fuel",count=5})
elseif ADD_CAR == "Train" then
	table.insert(kit,{name="locomotive",count=1})
	table.insert(kit,{name="cargo-wagon",count=2})
	table.insert(kit,{name="fluid-wagon",count=1})
	table.insert(kit,{name="solid-fuel",count=50})
	table.insert(kit,{name="rail",count=300})
	table.insert(kit,{name="train-stop",count=2})
	table.insert(kit,{name="rail-signal",count=20})
	table.insert(kit,{name="rail-chain-signal",count=20})
end

if PERFECTIONIST > 0 then
	table.insert(kit,{name="landfill",count=PERFECTIONIST})
end

-- MODS
--if settings.startup["start-with-vehicle-miner"].value == true then
--	table.insert(kit,{name="vehicle-miner",count=1})
--end
--
--if settings.startup["start-with-unit-remote-control"].value == true then
--	table.insert(kit,{name="unit-remote-control",count=1})
--end

-- Now lets get the player that just joined, and clear all of his inventory to replace with ours.
script.on_event(defines.events.on_player_created,function(param)
	local p=game.players[param.player_index]
	
	if START_EQUIP ~= "" then
		local items = split(START_EQUIP,",")
		for no, item_str in pairs(items) do
			local info = split(item_str,":")
			if game.item_prototypes[info[1]] then
				table.insert(kit, {name=info[1],count=tonumber(info[2])})
			end
		end
	end

	--[[if START_TECH == "everything" then
		--p.force.research_all_technologies()
	elseif ~= "" then
		local techs = split(START_TECH,",")
		for no, tech_str in pairs(techs) do
			--p.force.technology[techs].enabled=true
		end
	end]]--

	-- Here we run a check to clear all sections of his inventory.
	for i,v in pairs(defines.inventory) do
		pcall(function()
			p.get_inventory(v).clear()
		end)
	end

	-- Now we will add all the items from our mod settings choices.
	for i,v in pairs(kit) do
		p.insert(v)
	end
end)

config.lua
require "utils"

---------------------------------------------------------------------------
--                           INITIALIZATION                              --
---------------------------------------------------------------------------
AS = {}

-- Substation multiplier, default = 4
AS.multiplier = 4

-- Allow chests
AS.chests = 1

-- Vanilla configuratioins
AS.config = {
  ["basic-kit-mining"] = {
    ["None"] = {},
    ["Base"] = {
      {   1, "burner-mining-drill" },
      {   1, "stone-furnace" }
    },
    ["Light Industry"] = {
      {   2, "electric-mining-drill" },
      {   1, "medium-electric-pole" },
      {   5, "stone-furnace" }
    },
    ["Medium Industry"] = {
      {   4, "electric-mining-drill" },
      {   5, "medium-electric-pole" },
      {   4, "steel-furnace" }
    },
    ["Heavy Industry"] = {
      {  10, "electric-mining-drill" },
      {  10, "medium-electric-pole" },
      {  10, "steel-furnace" }
    },
    ["Tycoon"] = {
      {  19, "burner-mining-drill" },
      {  20, "electric-mining-drill" },
      {   4 * AS.multiplier, "substation" },
      {  20, "electric-furnace" }
    }
  },
  ["basic-kit-military"] = {
    ["None"] = {},
    ["Civilian"] = {
      {  10, "firearm-magazine"}
    },
    ["Soldier"] = {
      {   1, "submachine-gun" },
      {   1, "shotgun" },
      {  20, "piercing-rounds-magazine"},
      {  20, "piercing-shotgun-shell"}
    },
    ["Rambo"] = {
      {   1, "submachine-gun"},
      {   1, "flamethrower"},
      {   1, "combat-shotgun"},
      { 100, "uranium-rounds-magazine" },
      {  50, "flamethrower-ammo"},
      {  50, "piercing-shotgun-shell"}
    },
    ["Jedi"] = {
      {   1, "power-armor-mk2"},
      {   3, "fusion-reactor-equipment"},
      {   8, "energy-shield-mk2-equipment"},
      {  12, "personal-laser-defense-equipment" },
      {   5, "battery-mk2-equipment"},
      {   1, "spidertron"},
      {   3, "exoskeleton-equipment"}
    }
  },
  ["basic-kit-power"] = {
    ["None"] = {},
    ["Steam"] = {
      {   1, "offshore-pump"},
      {  10, "pipe"},
      {   1, "boiler"},
      {   1, "steam-engine"}
    },
    ["Advanced"] = {
      {   1, "offshore-pump"},
      {  10, "pipe"},
      {   1, "boiler"},
      {   1, "steam-engine"},
      {   6, "solar-panel" },
      {   3, "accumulator"},
      {   2 * AS.multiplier, "substation"}
    },
    ["Solar"] = {
      {  12, "solar-panel" },
      {   6, "accumulator"},
      {   4 * AS.multiplier, "substation"}
    },
    ["Solar Field"] = {
      {  50, "solar-panel" },
      {  20, "accumulator"},
      {   8 * AS.multiplier, "substation"}
    }
  },
  ["basic-kit-booster"] = {
    ["None"] = {},
    ["Resources"] = {
      {  50, "iron-plate" },
      {  50, "copper-plate"},
      {  50, "iron-gear-wheel"},
      {  20, "electronic-circuit"}
    },
    ["Power Armor"] = {
      {   1, "modular-armor" },
      {   1, "energy-shield-equipment" },
      {   1, "night-vision-equipment" },
      {   1, "battery-equipment" },
      {   7, "solar-panel-equipment"},
      {   1, "exoskeleton-equipment"}
    },
    ["Research Equipment"] = {
      {   1, "lab"},
      { 200, "automation-science-pack"}, --  TODO:timed technology?
      {  50, "logistic-science-pack"}
    },
    ["Power Armor & Resources"] = {
      {  50, "iron-plate" },
      {  50, "copper-plate"},
      {  50, "iron-gear-wheel"},
      {  20, "electronic-circuit"},
      {   1, "modular-armor" },
      {   1, "energy-shield-equipment" },
      {   1, "night-vision-equipment" },
      {   1, "battery-equipment" },
      {   7, "solar-panel-equipment"},
      {   1, "exoskeleton-equipment"}
    },
    ["All"] = {
      {  50, "iron-plate" },
      {  50, "copper-plate"},
      {  50, "iron-gear-wheel"},
      {  20, "electronic-circuit"},
      {   1, "modular-armor" },
      {   1, "energy-shield-equipment" },
      {   1, "night-vision-equipment" },
      {   1, "battery-equipment" },
      {   7, "solar-panel-equipment"},
      {   1, "exoskeleton-equipment"},
      {   1, "lab"},
      { 200, "automation-science-pack"},
      {  50, "logistic-science-pack" }
    }
  },
  ["basic-kit-modules"] = {
    ["None"] = {},
    ["Poker Speed"] = {
      {   4, "speed-module-3" }
    },
    ["Poker Productivity"] = {
      {   4, "productivity-module-3"}
    },
    ["Poker Efficiency"] = {
      {   4, "effectivity-module-3"}
    },
    ["Tris of Poker"] = {
      {   4, "speed-module-3" },
      {   4, "productivity-module-3"},
      {   4, "effectivity-module-3"}
    },
    ["Royal Scale"] = {
      {  16, "speed-module-3" },
      {  16, "productivity-module-3"},
      {  16, "effectivity-module-3"}
    }
  },
  ["basic-kit-robot"] = {
    ["None"] = {},
    ["Base"] = {
      {   1, "roboport" },
      {  10, "logistic-robot" },
      {  10, "construction-robot"},
      {   3 * AS.chests, "logistic-chest-passive-provider" },
      {   3 * AS.chests, "logistic-chest-storage" },
      {   1 * AS.chests, "logistic-chest-requester" }
    },
    ["Pack"] = {
      {   1, "roboport" },
      {  25, "logistic-robot" },
      {  25, "construction-robot"},
      {   6 * AS.chests, "logistic-chest-passive-provider" },
      {   6 * AS.chests, "logistic-chest-storage" },
      {   3 * AS.chests, "logistic-chest-requester" }
    },
    ["Flock"] = {
      {   2, "roboport" },
      {  50, "logistic-robot" },
      {  50, "construction-robot"},
      {  10 * AS.chests, "logistic-chest-passive-provider" },
      {  10 * AS.chests, "logistic-chest-storage" },
      {   5 * AS.chests, "logistic-chest-requester" }
    },
    ["Fleet"] = {
      {   4, "roboport" },
      { 100, "logistic-robot" },
      { 100, "construction-robot"},
      {  15 * AS.chests, "logistic-chest-passive-provider" },
      {  15 * AS.chests, "logistic-chest-storage" },
      {  10 * AS.chests, "logistic-chest-requester" },
      {   5 * AS.chests, "logistic-chest-active-provider" },
      {   5 * AS.chests, "logistic-chest-buffer" }
    }
  },
  ["basic-kit-vehicle"] = {
    ["None"] = {},
    ["Car"] = {
      {   1, "car" },
      {   5, "solid-fuel" }
    },
    ["Tank"] = {
      {   1, "tank" },
      {  10, "solid-fuel" }
    },
    ["Train"] = {
      {   1, "locomotive" },
      {   2, "cargo-wagon" },
      {   1, "fluid-wagon" },
      {  50, "solid-fuel" },
      { 300, "rail" },
      {   2, "train-stop" },
      {  20, "rail-signal" },
      {  20, "rail-chain-signal" }
    }
  },
  ["basic-kit-landfill"] = {
    ["None"] = {},
    ["Small"] = {
      {  50, "landfill" }
    },
    ["Medium"] = {
      { 100, "landfill" }
    },
    ["Large"] = {
      { 500, "landfill" }
    },
    ["Huge"] = {
      { 1000, "landfill" }
    }
  }
}

---------------------------------------------------------------------------
--                            COMPATIBILITY                              --
---------------------------------------------------------------------------

-- Bob's Modpack
if script.active_mods["boblogistics"] then
  RemoveAllItemOccurrences(AS.config, "roboport")

  local dir = AS.config["basic-kit-robot"]
  local cd = nil

  cd = dir["Base"]
  AddItem(cd, { 1, "bob-robochest" })
  AddItem(cd, { 1, "bob-robo-charge-port" })
  AddItem(cd, { 1, "bob-logistic-zone-interface"})
  AddItem(cd, { 1, "bob-logistic-zone-expander-2"})

  cd = dir["Pack"]
  AddItem(cd, { 1, "bob-robochest" })
  AddItem(cd, { 2, "bob-robo-charge-port" })
  AddItem(cd, { 1, "bob-logistic-zone-interface"})
  AddItem(cd, { 1, "bob-logistic-zone-expander"})

  cd = dir["Flock"]
  AddItem(cd, { 1, "bob-robochest" })
  AddItem(cd, { 3, "bob-robo-charge-port" })
  AddItem(cd, { 1, "bob-logistic-zone-interface"})
  AddItem(cd, { 2, "bob-logistic-zone-expander-3"})

  cd = dir["Fleet"]
  AddItem(cd, { 2, "bob-robochest" })
  AddItem(cd, { 4, "bob-robo-charge-port" })
  AddItem(cd, { 2, "bob-robo-charge-port-large" })
  AddItem(cd, { 1, "bob-logistic-zone-interface"})
  AddItem(cd, { 2, "bob-logistic-zone-expander-4"})
end

---------------------------------------------------------------------------
